connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java

```diff
@@ -102,6 +101,62 @@ public static boolean isJar(Path path) {
         return locations;
     }
 
+    public static List<Path> pluginUrls(Path topPath) throws IOException {
+        Set<Path> archives = new HashSet<>();
+        LinkedList<DirectoryEntry> dfs = new LinkedList<>();
+        Set<Path> visited = new HashSet<>();
+
+        if (isArchive(topPath)) {
+            return Collections.singletonList(topPath);
+        }
+
+        DirectoryStream<Path> topListing = Files.newDirectoryStream(
+                topPath,
+                PLUGIN_FILTER
+        );
+        dfs.push(new DirectoryEntry(topListing));
+        visited.add(topPath);
+        try {
+            while (!dfs.isEmpty()) {
+                Iterator<Path> neighbors = dfs.peek().iterator;
+                if (neighbors.hasNext()) {
```

Comment:
Not really a big deal, but the nesting in this method wouldn't get quite so bad if you switched some checks for some conditions and used `continue`. This `if`/`else` block is one example, another is where you set `adjacent = null` and then immediately check that condition.